### PR TITLE
jeepney: add trio server

### PR DESCRIPTION
**Pull request by takluyver**
Copy/paste the blocking backend, scatter some `async` and `await` around, and it basically seems to work. :-)

Tests equivalent to the blocking backend, using [pytest-trio](https://pytest-trio.readthedocs.io/en/stable/) to help write tests with Trio.

This doesn't yet know about async methods on the objects, though I don't think it would be hard to add. I think the backends could also share some code, in particular the bits of `_handle_msg` that deal with looking up the method and validating the signature. But I wanted to do it the easy way first before getting clever.

I'm not really asking you to merge this as is, it's more for discussion at the moment.

### CODE REVIEW

**Comment by takluyver** [comment]
Extra thought: once it supports async handler methods, it's also conceivable that you'd want multiple tasks handling messages from the same D-Bus connection. I think this should mostly work with several tasks calling `await server.listen()` on the same server object, but there might be some wrinkles with the auto-reset functionality.

**Review by FFY00 — CHANGES_REQUESTED** [review]
Thanks! Just some minor comments.

I think we should keep this in `dbus_objects.integration.jeepney`. This was what I originally had in mind, hence the `BlockingDBusServer` and not `DBusServer`, and it would make it easier to refactor out the Jeepney specific code into a base class.

If you have time, you cloud split the class in one commit and then add the trio class in another (`git add -p` is very useful there :stuck_out_tongue:). Otherwise, I can merge the unrefactored code and do that myself. Let me know what you want to do :blush:

So for now I just request that you put this in `jeepney.py` and use future annotations, as pointed out below.

**Comment by FFY00** [comment]
About support for async methods in objects, I think we just need to make `DBusSignature` unwrap the possible `typing.Awaitable` type in `from_return` and then the backend implementations should be able to use `inspect.iscoroutinefunction()` to see if they should call the method in async context or not.

**Comment by takluyver** [comment]
> I think we should keep this in dbus_objects.integration.jeepney. This was what I originally had in mind, hence the BlockingDBusServer and not DBusServer, and it would make it easier to refactor out the Jeepney specific code into a base class.

I wondered if that was what you were going for, but that means you would need trio installed in order to import the combined module, which doesn't seem right. Unless you prefer to hide all of the trio & jeepney.io.trio imports inside the methods that use them?

**Comment by FFY00** [comment]
I was thinking of having a soft-fail global import and raising an exception in `__init__` if the module is not available.

**Comment by FFY00** [comment]
@takluyver sorry for bothering, it seems the tests are failing with `trio.TooSlowError` (they also fail without my refactoring commit), do you have any idea why?

**Comment by takluyver** [comment]
Not sure. Does it come back if you put the `yield` back in the `jeepney_trio_one_time_server` fixture?

File: dbus_objects/integration/jeepney.py [hunk 1]
```diff
@@ -122,20 +151,92 @@ def listen(self, delay: float = 0.1, event: Optional[threading.Event] = None) ->
...
+        await inst._conn_start()
+        return inst
+
+    async def _conn_start(self) -> None:
+        '''
+        Start DBus connection
+        '''
+        self._conn = await jeepney.io.trio.open_dbus_connection(self._bus)
+        async with self._conn.router() as router:
+            bus_proxy = jeepney.io.trio.Proxy(jeepney.message_bus, router)
```
Comment (takluyver) [inline dbus_objects/integration/jeepney.py:199]:
```suggestion
        import jeepney.io.trio
        self._conn = await jeepney.io.trio.open_dbus_connection(self._bus)
        async with self._conn.router() as router:
            bus_proxy = jeepney.io.trio.Proxy(jeepney.message_bus, router)
```

Comment (FFY00) [reply dbus_objects/integration/jeepney.py:199]: Ah yes, I forgot about this.

**Comment by takluyver** [comment]
Testing locally, when I run only this test (`dbus-run-session pytest tests/ -k 'test_listen and trio'`), it consistently passes. When I run it along with the blocking test_listen (`dbus-run-session pytest tests/ -k 'test_listen and trio'`), it usually (but not always) fails. So it seems to be some interaction between the tests or their fixtures.

I wonder if when a process exits, there's a small delay before the bus daemon unregisters its connection and lets another connection take the name, so the test message for trio is getting routed to the defunct server for the blocking test. Maybe closing the connection explicitly before the process exits would help. It might also be a good idea to check the return value from calling `RequestName` ([docs](https://dbus.freedesktop.org/doc/dbus-specification.html#bus-messages-request-name)).

**Comment by FFY00** [comment]
I did notice that too. I tried using a different name and is still the same.

I added a few debug messages and registered the example tracer (https://trio.readthedocs.io/en/latest/tutorial.html#task-switching-illustrated) and I get the following output. Doesn't seem pytest-trio is running the listen task.

```
### task scheduled: <fixture 'jeepney_trio_server'>
<<< task step finished: <fixture 'jeepney_trio_server'>
>>> about to run one step of task: <fixture 'jeepney_trio_router'>
### task scheduled: <fixture 'jeepney_trio_router'>
<<< task step finished: <fixture 'jeepney_trio_router'>
### doing a quick check for I/O
### finished I/O check (took 9.67600499279797e-06 seconds)
>>> about to run one step of task: <fixture 'jeepney_trio_server'>
### task scheduled: <fixture 'jeepney_trio_server'>
<<< task step finished: <fixture 'jeepney_trio_server'>
>>> about to run one step of task: <fixture 'jeepney_trio_router'>
### task scheduled: <fixture 'jeepney_trio_router'>
<<< task step finished: <fixture 'jeepney_trio_router'>
### doing a quick check for I/O
### finished I/O check (took 7.86097371019423e-06 seconds)
>>> about to run one step of task: <fixture 'jeepney_trio_router'>
### task scheduled: <fixture 'jeepney_trio_router'>
<<< task step finished: <fixture 'jeepney_trio_router'>
>>> about to run one step of task: <fixture 'jeepney_trio_server'>
### task scheduled: <fixture 'jeepney_trio_server'>
<<< task step finished: <fixture 'jeepney_trio_server'>
### doing a quick check for I/O
### finished I/O check (took 7.84100848250091e-06 seconds)
>>> about to run one step of task: <fixture 'jeepney_trio_router'>
### new task spawned: jeepney.io.trio.DBusRouter._sender
### task scheduled: jeepney.io.trio.DBusRouter._sender
<<< task step finished: <fixture 'jeepney_trio_router'>
>>> about to run one step of task: <fixture 'jeepney_trio_server'>
### task scheduled: <fixture 'jeepney_trio_server'>
<<< task step finished: <fixture 'jeepney_trio_server'>
### doing a quick check for I/O
### finished I/O check (took 7.924973033368587e-06 seconds)
>>> about to run one step of task: <fixture 'jeepney_trio_server'>
### new task spawned: jeepney.io.trio.DBusRouter._sender
### task scheduled: jeepney.io.trio.DBusRouter._sender
<<< task step finished: <fixture 'jeepney_trio_server'>
>>> about to run one step of task: jeepney.io.trio.DBusRouter._sender
### task scheduled: <fixture 'jeepney_trio_router'>
<<< task step finished: jeepney.io.trio.DBusRouter._sender
### doing a quick check for I/O
### finished I/O check (took 7.601018296554685e-06 seconds)
>>> about to run one step of task: <fixture 'jeepney_trio_router'>
### new task spawned: jeepney.io.trio.DBusRouter._receiver
### task scheduled: jeepney.io.trio.DBusRouter._receiver
<<< task step finished: <fixture 'jeepney_trio_router'>
>>> about to run one step of task: jeepney.io.trio.DBusRouter._sender
### task scheduled: <fixture 'jeepney_trio_server'>
<<< task step finished: jeepney.io.trio.DBusRouter._sender
### doing a quick check for I/O
### finished I/O check (took 7.488008122891188e-06 seconds)
>>> about to run one step of task: jeepney.io.trio.DBusRouter._receiver
### task scheduled: <fixture 'jeepney_trio_router'>
<<< task step finished: jeepney.io.trio.DBusRouter._receiver
>>> about to run one step of task: <fixture 'jeepney_trio_server'>
### new task spawned: jeepney.io.trio.DBusRouter._receiver
### task scheduled: jeepney.io.trio.DBusRouter._receiver
<<< task step finished: <fixture 'jeepney_trio_server'>
### doing a quick check for I/O
### finished I/O check (took 7.658964022994041e-06 seconds)
>>> about to run one step of task: jeepney.io.trio.DBusRouter._receiver
### task scheduled: <fixture 'jeepney_trio_server'>
<<< task step finished: jeepney.io.trio.DBusRouter._receiver
>>> about to run one step of task: <fixture 'jeepney_trio_router'>
### task scheduled: jeepney.io.trio.DBusRouter._sender
### task scheduled: <fixture 'jeepney_trio_router'>
<<< task step finished: <fixture 'jeepney_trio_router'>
### doing a quick check for I/O
### finished I/O check (took 7.7290169429034e-06 seconds)
>>> about to run one step of task: <fixture 'jeepney_trio_server'>
### task scheduled: jeepney.io.trio.DBusRouter._sender
### task scheduled: <fixture 'jeepney_trio_server'>
<<< task step finished: <fixture 'jeepney_trio_server'>
>>> about to run one step of task: jeepney.io.trio.DBusRouter._sender
### task scheduled: jeepney.io.trio.DBusRouter._sender
<<< task step finished: jeepney.io.trio.DBusRouter._sender
>>> about to run one step of task: <fixture 'jeepney_trio_router'>
<<< task step finished: <fixture 'jeepney_trio_router'>
### doing a quick check for I/O
### finished I/O check (took 7.681024726480246e-06 seconds)
>>> about to run one step of task: jeepney.io.trio.DBusRouter._sender
### task scheduled: jeepney.io.trio.DBusRouter._sender
<<< task step finished: jeepney.io.trio.DBusRouter._sender
>>> about to run one step of task: <fixture 'jeepney_trio_server'>
<<< task step finished: <fixture 'jeepney_trio_server'>
>>> about to run one step of task: jeepney.io.trio.DBusRouter._sender
### task scheduled: jeepney.io.trio.DBusRouter._sender
<<< task step finished: jeepney.io.trio.DBusRouter._sender
### doing a quick check for I/O
### finished I/O check (took 1.259200507774949e-05 seconds)
>>> about to run one step of task: jeepney.io.trio.DBusRouter._sender
### task scheduled: jeepney.io.trio.DBusRouter._sender
<<< task step finished: jeepney.io.trio.DBusRouter._sender
>>> about to run one step of task: jeepney.io.trio.DBusRouter._sender
<<< task step finished: jeepney.io.trio.DBusRouter._sender
### doing a quick check for I/O
### task scheduled: jeepney.io.trio.DBusRouter._receiver
### finished I/O check (took 2.328603295609355e-05 seconds)
>>> about to run one step of task: jeepney.io.trio.DBusRouter._receiver
### task scheduled: <fixture 'jeepney_trio_router'>
<<< task step finished: jeepney.io.trio.DBusRouter._receiver
>>> about to run one step of task: jeepney.io.trio.DBusRouter._sender
<<< task step finished: jeepney.io.trio.DBusRouter._sender
### doing a quick check for I/O
### task scheduled: jeepney.io.trio.DBusRouter._receiver
### finished I/O check (took 3.1830015359446406e-05 seconds)
>>> about to run one step of task: jeepney.io.trio.DBusRouter._receiver
### task scheduled: <fixture 'jeepney_trio_server'>
<<< task step finished: jeepney.io.trio.DBusRouter._receiver
>>> about to run one step of task: <fixture 'jeepney_trio_router'>
### task scheduled: jeepney.io.trio.DBusRouter._sender
### task scheduled: <fixture 'jeepney_trio_router'>
<<< task step finished: <fixture 'jeepney_trio_router'>
### doing a quick check for I/O
### finished I/O check (took 1.0684016160666943e-05 seconds)
>>> about to run one step of task: <fixture 'jeepney_trio_server'>
### task scheduled: jeepney.io.trio.DBusRouter._sender
### task scheduled: <fixture 'jeepney_trio_server'>
<<< task step finished: <fixture 'jeepney_trio_server'>
>>> about to run one step of task: jeepney.io.trio.DBusRouter._sender
### task scheduled: jeepney.io.trio.DBusRouter._sender
<<< task step finished: jeepney.io.trio.DBusRouter._sender
>>> about to run one step of task: <fixture 'jeepney_trio_router'>
### task scheduled: jeepney.io.trio.DBusRouter._receiver
### task scheduled: <fixture 'jeepney_trio_router'>
<<< task step finished: <fixture 'jeepney_trio_router'>
### doing a quick check for I/O
### finished I/O check (took 1.0647985618561506e-05 seconds)
>>> about to run one step of task: <fixture 'jeepney_trio_router'>
<<< task step finished: <fixture 'jeepney_trio_router'>
>>> about to run one step of task: jeepney.io.trio.DBusRouter._receiver
### task exited: jeepney.io.trio.DBusRouter._receiver
<<< task step finished: jeepney.io.trio.DBusRouter._receiver
>>> about to run one step of task: jeepney.io.trio.DBusRouter._sender
### task scheduled: <fixture 'jeepney_trio_router'>
### task exited: jeepney.io.trio.DBusRouter._sender
<<< task step finished: jeepney.io.trio.DBusRouter._sender
>>> about to run one step of task: <fixture 'jeepney_trio_server'>
### task scheduled: jeepney.io.trio.DBusRouter._receiver
### task scheduled: <fixture 'jeepney_trio_server'>
<<< task step finished: <fixture 'jeepney_trio_server'>
>>> about to run one step of task: jeepney.io.trio.DBusRouter._sender
### task scheduled: jeepney.io.trio.DBusRouter._sender
<<< task step finished: jeepney.io.trio.DBusRouter._sender
### doing a quick check for I/O
### finished I/O check (took 1.0263989679515362e-05 seconds)
>>> about to run one step of task: jeepney.io.trio.DBusRouter._sender
### task exited: jeepney.io.trio.DBusRouter._sender
<<< task step finished: jeepney.io.trio.DBusRouter._sender
>>> about to run one step of task: <fixture 'jeepney_trio_server'>
<<< task step finished: <fixture 'jeepney_trio_server'>
>>> about to run one step of task: jeepney.io.trio.DBusRouter._receiver
### task scheduled: <fixture 'jeepney_trio_server'>
### task exited: jeepney.io.trio.DBusRouter._receiver
<<< task step finished: jeepney.io.trio.DBusRouter._receiver
>>> about to run one step of task: <fixture 'jeepney_trio_router'>
### new task spawned: jeepney.io.trio.DBusRouter._sender
### task scheduled: jeepney.io.trio.DBusRouter._sender
<<< task step finished: <fixture 'jeepney_trio_router'>
### doing a quick check for I/O
### finished I/O check (took 1.777301076799631e-05 seconds)
>>> about to run one step of task: jeepney.io.trio.DBusRouter._sender
### task scheduled: <fixture 'jeepney_trio_router'>
<<< task step finished: jeepney.io.trio.DBusRouter._sender
>>> about to run one step of task: <fixture 'jeepney_trio_server'>
### new task spawned: jeepney.io.trio.DBusRouter._sender
### task scheduled: jeepney.io.trio.DBusRouter._sender
<<< task step finished: <fixture 'jeepney_trio_server'>
### doing a quick check for I/O
### finished I/O check (took 8.350994903594255e-06 seconds)
>>> about to run one step of task: <fixture 'jeepney_trio_router'>
### new task spawned: jeepney.io.trio.DBusRouter._receiver
### task scheduled: jeepney.io.trio.DBusRouter._receiver
<<< task step finished: <fixture 'jeepney_trio_router'>
>>> about to run one step of task: jeepney.io.trio.DBusRouter._sender
### task scheduled: <fixture 'jeepney_trio_server'>
<<< task step finished: jeepney.io.trio.DBusRouter._sender
### doing a quick check for I/O
### finished I/O check (took 1.0610994650050998e-05 seconds)
>>> about to run one step of task: <fixture 'jeepney_trio_server'>
### new task spawned: jeepney.io.trio.DBusRouter._receiver
### task scheduled: jeepney.io.trio.DBusRouter._receiver
<<< task step finished: <fixture 'jeepney_trio_server'>
>>> about to run one step of task: jeepney.io.trio.DBusRouter._receiver
### task scheduled: <fixture 'jeepney_trio_router'>
<<< task step finished: jeepney.io.trio.DBusRouter._receiver
### doing a quick check for I/O
### finished I/O check (took 1.0144984116777778e-05 seconds)
>>> about to run one step of task: jeepney.io.trio.DBusRouter._receiver
### task scheduled: <fixture 'jeepney_trio_server'>
<<< task step finished: jeepney.io.trio.DBusRouter._receiver
>>> about to run one step of task: <fixture 'jeepney_trio_router'>
### task scheduled: <test 'test_listen_trio'>
<<< task step finished: <fixture 'jeepney_trio_router'>
### doing a quick check for I/O
### finished I/O check (took 1.0098970960825682e-05 seconds)
>>> about to run one step of task: <fixture 'jeepney_trio_server'>
### task scheduled: jeepney.io.trio.DBusRouter._sender
### task scheduled: <fixture 'jeepney_trio_server'>
<<< task step finished: <fixture 'jeepney_trio_server'>
>>> about to run one step of task: <test 'test_listen_trio'>
<<< task step finished: <test 'test_listen_trio'>
### doing a quick check for I/O
### finished I/O check (took 1.0446005035191774e-05 seconds)
>>> about to run one step of task: jeepney.io.trio.DBusRouter._sender
### task scheduled: jeepney.io.trio.DBusRouter._sender
<<< task step finished: jeepney.io.trio.DBusRouter._sender
>>> about to run one step of task: <fixture 'jeepney_trio_server'>
<<< task step finished: <fixture 'jeepney_trio_server'>
### doing a quick check for I/O
### finished I/O check (took 1.0140996892005205e-05 seconds)
>>> about to run one step of task: jeepney.io.trio.DBusRouter._sender
### task scheduled: jeepney.io.trio.DBusRouter._sender
<<< task step finished: jeepney.io.trio.DBusRouter._sender
### doing a quick check for I/O
### finished I/O check (took 9.871990187093616e-06 seconds)
>>> about to run one step of task: jeepney.io.trio.DBusRouter._sender
<<< task step finished: jeepney.io.trio.DBusRouter._sender
### waiting for I/O for up to 86400 seconds
### task scheduled: jeepney.io.trio.DBusRouter._receiver
### finished I/O check (took 3.1194009352475405e-05 seconds)
>>> about to run one step of task: jeepney.io.trio.DBusRouter._receiver
### task scheduled: <fixture 'jeepney_trio_server'>
<<< task step finished: jeepney.io.trio.DBusRouter._receiver
### doing a quick check for I/O
### finished I/O check (took 1.096399500966072e-05 seconds)
>>> about to run one step of task: <fixture 'jeepney_trio_server'>
### task scheduled: jeepney.io.trio.DBusRouter._sender
### task scheduled: <fixture 'jeepney_trio_server'>
<<< task step finished: <fixture 'jeepney_trio_server'>
### doing a quick check for I/O
### finished I/O check (took 1.02429767139256e-05 seconds)
>>> about to run one step of task: <fixture 'jeepney_trio_server'>
### task scheduled: jeepney.io.trio.DBusRouter._receiver
### task scheduled: <fixture 'jeepney_trio_server'>
<<< task step finished: <fixture 'jeepney_trio_server'>
>>> about to run one step of task: jeepney.io.trio.DBusRouter._sender
### task scheduled: jeepney.io.trio.DBusRouter._sender
<<< task step finished: jeepney.io.trio.DBusRouter._sender
### doing a quick check for I/O
### finished I/O check (took 9.790994226932526e-06 seconds)
>>> about to run one step of task: jeepney.io.trio.DBusRouter._receiver
### task exited: jeepney.io.trio.DBusRouter._receiver
<<< task step finished: jeepney.io.trio.DBusRouter._receiver
>>> about to run one step of task: <fixture 'jeepney_trio_server'>
<<< task step finished: <fixture 'jeepney_trio_server'>
>>> about to run one step of task: jeepney.io.trio.DBusRouter._sender
### task scheduled: <fixture 'jeepney_trio_server'>
### task exited: jeepney.io.trio.DBusRouter._sender
<<< task step finished: jeepney.io.trio.DBusRouter._sender
### doing a quick check for I/O
### finished I/O check (took 1.0171002941206098e-05 seconds)
>>> about to run one step of task: <fixture 'jeepney_trio_server'>
ready!
### task scheduled: <test 'test_listen_trio'>
<<< task step finished: <fixture 'jeepney_trio_server'>
### doing a quick check for I/O
### finished I/O check (took 1.1056021321564913e-05 seconds)
>>> about to run one step of task: <test 'test_listen_trio'>
### task scheduled: <test 'test_listen_trio'>
<<< task step finished: <test 'test_listen_trio'>
### doing a quick check for I/O
### finished I/O check (took 1.0229006875306368e-05 seconds)
>>> about to run one step of task: <test 'test_listen_trio'>
### new task spawned: dbus_objects.integration.jeepney.TrioDBusServer.listen
### task scheduled: dbus_objects.integration.jeepney.TrioDBusServer.listen
sending...
### task scheduled: jeepney.io.trio.DBusRouter._sender
### task scheduled: <test 'test_listen_trio'>
<<< task step finished: <test 'test_listen_trio'>
### doing a quick check for I/O
### finished I/O check (took 2.125598257407546e-05 seconds)
>>> about to run one step of task: <test 'test_listen_trio'>
<<< task step finished: <test 'test_listen_trio'>
>>> about to run one step of task: jeepney.io.trio.DBusRouter._sender
### task scheduled: jeepney.io.trio.DBusRouter._sender
<<< task step finished: jeepney.io.trio.DBusRouter._sender
>>> about to run one step of task: dbus_objects.integration.jeepney.TrioDBusServer.listen
listening!
<<< task step finished: dbus_objects.integration.jeepney.TrioDBusServer.listen
### doing a quick check for I/O
### finished I/O check (took 1.0999006917700171e-05 seconds)
>>> about to run one step of task: jeepney.io.trio.DBusRouter._sender
### task scheduled: jeepney.io.trio.DBusRouter._sender
<<< task step finished: jeepney.io.trio.DBusRouter._sender
### doing a quick check for I/O
### finished I/O check (took 9.86401573754847e-06 seconds)
>>> about to run one step of task: jeepney.io.trio.DBusRouter._sender
<<< task step finished: jeepney.io.trio.DBusRouter._sender
### waiting for I/O for up to 86400 seconds
```

**Comment by FFY00** [comment]
Yeah, looks like the code blocks without scheduling the task...

**Comment by FFY00** [comment]
Okay, it was an issue with the DBus server tearing down the server from the other test, which uses the same name.

**Pull request merged**